Rewrite and extend computer graphics notes to follow a more conventional style (#3)
VIK-CE-Notes/vik-ce-notes.github.io: docs/notes/sem4/computer_graphics/lectures/6.md

## Alapok

Egy kétdimenziós területre vagy egy háromdimenziós felületre szeretnénk egy képet, vagy más szóval tapétát ragasztani. Ez a kép meghatározná a terület vagy felület színét, megjelenését, így akár komplex kinézetű objektumokat és meg tudunk jeleníteni.

Ezt a felületet meg kell feleltetnünk annak a koordináta-rendszernek, amelyben a képet ábrázoljuk. Ezt a koordináta-rendszert _textúra koordináta-rendszernek_ nevezzük, és a $(0,0)$, $(1,1)$ tartományban értelmezzük.

!!! info 6.1 Definíció (Paraméterezés)
    Azt a folyamatot, amikor a felület egyes pontjait leképezzük ebbe az egységnégyzetbe (textúra térbe) _paraméterezésnek_ nevezzük.

Ha a felületünk összes pontja le van képezve ebbe az egységnégyzetbe, és mi "kifestjük" ezt a négyzetet azzal a mintázattal, amit a felületünkre szeretnénk rakni, akkor az eredeti felületünk ki lesz színezve. Ehhez persze szükségünk lesz a paraméterezés _inverzéhez_, azaz ahhoz a transzformációhoz, ami a textúra koordináta-rendszer egyes pontjait megfelelteti a geometriánk felületének egyes pontjaival.

A textúrázáshoz használt kép elemei a _texelek_ (mint ahogy a pixel a "picture element"-ből származik, a texel a "texture element"-ből).

> A textúrázást elég csak háromszögekre megvalósítani, hiszen az összes többi bonyolultabb felületet háromszögekből építjük fel. Az a célunk, hogy a textúraterünk háromszögeit valahogy megfeleltessük a modellterünk háromszögeivel. Az egyik ilyen megfeleltetés, amit a leggyakrabban használunk az a _lineáris leképzés_.

Reviewer: ```suggestion
A textúrázást elég csak háromszögekre megvalósítani, hiszen az összes többi bonyolultabb felületet háromszögekből építjük fel. Az a célunk, hogy a textúraterünk háromszögeit valahogy megfeleltessük a modellterünk háromszögeivel. A leggyakrabban használt egyik ilyen megfeleltetés a _lineáris leképzés_.
```

## Lineáris leképzés

> A fő ötlet az, hogy _"bármely két háromszög között létezik olyan affin transzformáció, ami az egyik háromszöget a másikba viszi át"_ és ez az affin transzformáció Descartes koordinátákban kifejezve egy _lineáris kifejezés_, azaz mind $u$, mind $v$ elsőfokú hatványokként jelennek meg. Más szóval az a megfeleltetést keressük, ami az $(u,v)$ textúra térbeli koordinátákat megfelelteti a modell térbeli $(x,y,z)$ koordinátáknak, és a kapcsolatuk lineáris, azaz:

Reviewer: ```suggestion
Bármely két háromszög között létezik olyan affin transzformáció, ami az egyik háromszöget a másikba viszi át. Ez az affin transzformáció Descartes koordinátákban kifejezve egy _lineáris kifejezés_, azaz $u$ és $v$ elsőfokú hatványokként szerepelnek. Tehát azt a megfeleltetést keressük, ami az $(u,v)$ textúra térbeli koordinátákat megfelelteti a modell térbeli $(x,y,z)$ koordinátáknak, és a kapcsolatuk lineáris, azaz:
```

\end{equation}
$$

### Geometriai megközelítés

Annak, hogy $U$ az $(x,y)$ koordináták lineáris függvénye más következményei is vannak. Ha felveszünk egy olyan háromdimenziós koordináta-rendszert, ahol a tengelyek rendre $x,y,U$, akkor ebben a koordináta-rendszerben azok az $(x, y, U)$ számhármasok, amik a háromszögeink adatait tartalmazzák egy síkra kerülnek. Ez látható a sík képletéből is:

$$
\begin{alignat*}{2}
\text{Sík egyenlete:} & \quad n_x \cdot x + n_y \cdot y + n_u \cdot U + d = 0 \\
\text{Interpolációs feltétel:} & \quad U(x,y) = A_u \cdot x + B_u \cdot y + C_u \\[1.5ex]
\text{Átrendezett sík egyenlet:} & \quad U = -\frac{n_x}{n_u} \cdot x - \frac{n_y}{n_u} \cdot y - \frac{d}{n_u}
\end{alignat*}
$$

Tehát nekünk $\displaystyle -\frac{n_x}{n_u}$ alakban kell az eredmény, mert az az $A_u$ megfelelője az interpolációs feltételben. A sík normálvektorát megkaphatjuk úgy, ha a síkban két nem párhuzamos vektornak vesszük a keresztszorzatát. Mivel a háromszögünk ebben a síkban van, ezért már három pontot ismerünk a síkból. Ebből a három ponton könnyen tudunk két alkalmas vektort képezni, és ezek keresztszorzata pontosan megegyezik a fenti $(1)$ képlettel.

## Textúra szűrés

> Eljutottunk odáig, hogy minden egyes pixelt hatékonyan át tudunk transzformálni a textúra térbe. Viszont [nem triviális](https://cdn.discordapp.com/emojis/1384476146636161065.webp?size=240), hogy ez alapján a transzformált pont alapján melyik texelt válasszuk ki. Ez a kiválasztási folyamat a _textúra szűrés_. E során a pixelek középpontjait transzformáljuk a textúra térbe, és megnézzük, hogy melyik texel középpontjához vannak a legközelebb. Ez a `GL_NEAREST`, más néven _matematikai mintavételezés_. Ez alap esetben jól működik, viszont két olyan eset mégis felmerülhet, amikor nekünk ez nem megfelelő:

Reviewer: ```suggestion
Eljutottunk odáig, hogy minden egyes pixelt hatékonyan át tudunk transzformálni a textúra térbe. Viszont [nem triviális](https://cdn.discordapp.com/emojis/1384476146636161065.webp?size=240), hogy ez alapján a transzformált pont alapján melyik texelt választjuk ki. Ez a kiválasztási folyamat a _textúra szűrés_. Ennek során a pixelek középpontjait transzformáljuk a textúra térbe, és megnézzük, hogy melyik texel középpontjához vannak a legközelebb. Ez a `GL_NEAREST` (nearest neighbor), más néven _matematikai mintavételezés_. Ez alapesetben jól működik, viszont két olyan eset is felmerülhet, amikor nekünk ez nem megfelelő:
```
